core/src/solution_submission.rs

```diff
@@ -238,15 +259,40 @@ async fn retry_with_gas_price_increase(
         } else {
             Some(BLOCK_TIMEOUT)
         };
-        let result = submit_solution(gas_price, block_timeout).await;
-        if !is_confirm_timeout(&result) || is_last_iteration {
-            return result;
+        let solution_future = submit_solution(gas_price, block_timeout);
+        let deadline_future = wait_for_deadline();
+        futures::pin_mut!(deadline_future);
+        match future::select(solution_future, deadline_future).await {
+            Either::Left((result, _)) => {
+                if !is_confirm_timeout(&result) || is_last_iteration {
+                    return result;
+                }
+                gas_price_factor *= INCREASE_FACTOR;
+                log::info!(
+                    "retrying solution submission with increased gas price factor of {}",
+                    gas_price_factor,
+                );
+            }
+            Either::Right(_) => {
+                let gas_price = U256::from(
+                    (gas_price.as_u128() as f64 * MIN_GAS_PRICE_INCREASE_FACTOR).ceil() as u128,
+                );
+                match contract.send_noop_transaction(gas_price).await {
+                    Ok(_) => log::info!(
+                        "cancel solution submission of batch {} because of deadline",
+                        batch_index
+                    ),
+                    Err(err) => log::error!(
+                        "failed to cancel solution submission of batch {} after deadline: {:?}",
+                        batch_index,
+                        err
+                    ),
+                }
+                // TODO: What to return? Should we make up a method error? Or we could make this
+                // function return Result<(), Either<MethodError, DeadlineReached>> .
```

Comment:
Equivalent to your suggestion, I think the cleanest would be our own SolutionSubmissionError enum which can itself be either MethodError or DeadlineReached